File: Python/remove_special_char.py
import os
import os.path

#TODO:
#Replace all special characters with the desired character
#This program assumes that file has an extension name

#Check if the character is an alphabet
def isAlpha (s):
    if ( ('a' <= s and s <= 'z') or ('A' <= s and s <= 'Z')):
        return True
    else:
        return False

#Fetch the last dot
def lastFullStop (fileName):
    end = len(fileName) - 1
    while (end >= 0):
        if (fileName[end] == '.'):
            return end
        end -=1
    #did not find last dot
    #This may be a bug
    if (end == -1):
        return 0

def main():
    #Fetch the desired character
    print("Enter a desired character")
    myChar = input()

    if (len(myChar) != 1): 
        print("Must only enter one character")
        exit()


    #Get a directory from the user
    print("Enter a directory")
    myDir = input()

    dir = os.scandir(myDir)
    #A temporary string variable to form a list
    joinString = ""
    #Check for alphabet
    for file in dir:
        fileName = list(file.name)
        oldFileName = file.name
        #replace all non-alphabet characters
        i = 0
        #Find the index of the last dot
        while(i < len(fileName) and i < lastFullStop(fileName)):
            s = fileName[i]
            if (not isAlpha(s)):
                fileName[i] = myChar
            i +=1
        fileName = joinString.join(fileName)
        print(fileName)
        dirPath = os.getcwd() + "/" + myDir +  "/"
        os.renames(oldFileName, fileName)

    #Do not delete the extension name
    #save resources
    os.scandir(myDir)

#execute main down here
if __name__ == "__main__":
    main()

    

File: C/removeSpecialChar.c
//A program that reads in a folder, and removes all special characters 
//and replaces them with an underscore or hyphen
//This program assumes that a file extension exists
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/types.h>
#include <ctype.h>

#define MAX_LEN 0xffff

//A function that gets the index of the last '.' for the file extension
int lastDotFileExtension (char *s) {
    int i = strlen(s);
    while (i > 0 && s[i] != ':') {
         i -=1;
    }
    return i;
}

//A function that replaces the special character
void replaceSpecialChar(DIR *d, char *parent, char replacement) {
    struct dirent *curr;
    while ((curr = readdir(d)) != NULL) {
        //Fetch the path of the file;
        char *currPath = calloc(MAX_LEN, sizeof(char));
        currPath = strcpy(currPath, parent);
        currPath = strcat(currPath, curr->d_name);

        //Fetch the position of dot of the file extension
        int lastDot = lastDotFileExtension(curr->d_name);
        char *renamedFile = calloc(MAX_LEN, sizeof(char));
        renamedFile = strcpy(renamedFile, curr->d_name);
        int i = 0;
        while (i < lastDot) {
            //Check for special characters
            if (!isalpha( (int)renamedFile[i] ) ) {
                //replace with desired characters
                renamedFile[i] = replacement;
            }
            i +=1;
        }
        char *newPath = calloc(MAX_LEN, sizeof(char) );
        newPath = strcpy(newPath, parent);
        //Construct a path with the new file name
        newPath = strcat(newPath, renamedFile);

        //Rename the file
        rename(currPath, newPath);

        //Free extra memory
        free(currPath);
        free(newPath);
        free(renamedFile);
    }
    free(parent);
}

//A function that reads a directory of files
int main (int argc, char **argv) {
    if (argc != 2) {
        printf("Incorrect number of arguments");
        exit(1);
    }
    //Fetch the directory pointer
    DIR *d = opendir(argv[1]);
    if (!d) {
        printf("Not a directory");
        exit(1);
    }
    
    //Change the replacement character here.
    char *parent = calloc(MAX_LEN, sizeof(char) );
    replaceSpecialChar(d, strcat(strcpy(parent, argv[1]), "/"), '_' );
    closedir(d);
}
